Pool_Properties_Design.md

```diff
@@ -0,0 +1,318 @@
+	// In the case that IsIsolated = false, failure of the domain is not a
+	// consideration.
+	RecoveryPointObjectiveTime string `json:"recoveryPointObjectiveTime,omitempty" yaml:"recoveryPointObjectiveTime,omitempty"`
+
+	// The enumeration literal specifies the time after a disaster that the
+	// client shall regain conformant service level access to the primary
+	// store. The possible values of this property could be:
+	// * OnlineActive: Active access to synchronous replicas.
+	// * OnlinePassive: Passive access to replicas via the same front-end
+	//   interconnect.
+	// * Nearline: Access to replica via a different front-end interconnect. A
+	//   restore step is required before recovery can commence.
+	// * Offline: No direct connection to the replica. (i.e. To a bunker
+	//   containing backup media.)
+	RecoveryTimeObjective string `json:"recoveryTimeObjective,omitempty" yaml:"recoveryTimeObjective,omitempty"`
+
+	// The enumeration literals may be used to specify the intended outcome of
+	// the replication. The possible values of this property could be:
+	// * Clone: This enumeration literal shall indicate that replication shall
+	//   create a point in time, full copy the source.
+	// * Mirror: This enumeration literal shall indicate that replication shall
+	//   create and maintain a copy of the source.
+	// * Snapshot: This enumeration literal shall indicate that replication
+	//   shall create a point in time, virtual copy of the source.
+	// * TokenizedClone: This enumeration literal shall indicate that replication
+	//   shall create a token based clone.
+	ReplicaType string `json:"replicaType,omitempty" yaml:"replicaType,omitempty"`
+}
+```
+
+* StoragePoolExtraSpec
+```go
+type StoragePoolExtraSpec struct {
+	// DataStorage represents some suggested data storage capabilities.
+	DataStorage DataStorageLoS `json:"dataStorage,omitempty" yaml:"dataStorage,omitempty"`
+	// IOConnectivity represents some suggested IO connectivity capabilities.
+	IOConnectivity IOConnectivityLoS `json:"ioConnectivity,omitempty" yaml:"ioConnectivity,omitempty"`
+	// DataProtection represents some suggested data protection capabilities.
+	DataProtection DataProtectionLos `json:"dataProtection,omitempty" yaml:"dataProtection,omitempty"`
+
+	// Besides those basic suggested pool properties, vendors can configure
+	// some advanced features (such as diskType, latency, compression, etc)
+	// themselves, all these properties can be exposed to controller scheduler
+	// and filtered by selector in a extensible way.
+	Advanced map[string]interface{} `json:"advanced,omitempty" yaml:"advanced,omitempty"`
+}
+```
+
+### REST API impact
+
+None
+
+### Security impact
+
+None
+
+### Other end user impact
+
+None
+
+### Performance impact
+
+Compared with before, there would be some performance loss when encoding and
+decoding the StoragePoolSpec. But we can improve the performance like this:
+```go
+type StoragePoolExtraSpec struct {
```

Comment:
How to express multiple capabilities in same xxxLoS such as a pool has both Snapshot and Clone capabilities on ReplicaType of DataProtection?

Reply:
Normally we suggest that the backend storage pools are atomic, which means it should not contain multiple capabilities in a single ***LoS. That's a solution for reducing the complexity of mapping profile with pool.
But if the backend wants to enable multiple capabilities, I think spliting into several atomic pools would be a better idea.

Reply:
Snapshot is a basic capability for each pool, therefore the ReplicaType of DataProtectionLoS will be set to snapshot, in that case the Clone, Mirror, etc. can't be expressed by using  ReplicaType of DataProtectionLoS.

IMO, there are a number of capabilites that belong to the same xxxLoS so that a atomic pool will be splited into several(two or more) atomic pools.